### PR TITLE
Add tooltips for MD editor buttons and add `muted` class for buttons

### PR DESCRIPTION
**Pull request by delvh**
Followup of #23876 according to my unreleased review demanding tooltips.
Additionally
- add a `muted` equivalent for buttons
- convert `switch to legacy` to an actual button
- enroll `switch to legacy` in the builtin pseudo focus cycle
- remove spaces between the buttons

The effect of the `muted` class is what you would expect: The button loses all of its normal styling, and is defined only by its content instead.
This will help reduce a11y infractions in the future, as that was one of the major points why people didn't use `<button>` tags and decided on a bad fix (i.e. through `<div>`s) instead.

## Appearance
image

### CODE REVIEW

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [inline templates/shared/combomarkdowneditor.tmpl]: If you make the parent a flexbox (`gt-df`), the whitespace between tags will not matter anymore.

File: web_src/css/base.css
Comment (wxiaoguang) [inline web_src/css/base.css]: `button.muted` doesn't mean the same as `a.muted`?

`.muted-buttons` is never used, and I guess it wouldn't help much in the future, remove it?

File: options/locale/locale_en-US.ini
Comment (wxiaoguang) [inline options/locale/locale_en-US.ini]: `header` doesn't mean `Add heading` in English?

File: options/locale/locale_en-US.ini
Comment (wxiaoguang) [inline options/locale/locale_en-US.ini]: I guess `EasyMDE` is clearer than `legacy editor` ?

File: web_src/css/base.css
Comment (silverwind) [reply web_src/css/base.css]: `.muted-buttons` was in use once where the button HTML were in the translation files so I could not directly alter them without invalidating the translations, but I think it's obsolete already and can be removed.

File: templates/shared/combomarkdowneditor.tmpl
Comment (wxiaoguang) [inline templates/shared/combomarkdowneditor.tmpl]: What's the `data-md-button` used for?

File: options/locale/locale_en-US.ini
Comment (silverwind) [reply options/locale/locale_en-US.ini]: I think legacy editor is fine. Users don't know what EasyMDE is.

File: web_src/css/base.css
Comment (wxiaoguang) [reply web_src/css/base.css]: In history, such design does cause problems.

For example:

```html
<div class="muted-buttons">
    <button ...>
    <div class="another-pannel">
        <button do-not-want-muted>
    </div>
</div>
```

The real case is #23888 ..... Fomantic UI used `:not()` selector to make `<button do-not-want-muted>` happy, but it's really a mess.

Comment (silverwind) [reply web_src/css/base.css]: Yes, I'm a strong supporter of eliminating HTML tags in translations.

File: templates/shared/combomarkdowneditor.tmpl
Comment (delvh) [reply templates/shared/combomarkdowneditor.tmpl]: > enroll `switch to legacy` in the builtin pseudo focus cycle

> https://github.com/github/markdown-toolbar-element#usage

Comment (wxiaoguang) [reply templates/shared/combomarkdowneditor.tmpl]: Good catch, I missed that.

File: web_src/css/base.css
Comment (delvh) [reply web_src/css/base.css]:
```suggestion
button.muted {
```

Comment (wxiaoguang) [reply web_src/css/base.css]: Do we have a definition of "muted"? What is a muted `a` or a muted `button`?

File: options/locale/locale_en-US.ini
Comment (wxiaoguang) [reply options/locale/locale_en-US.ini]: I guess the key could be something like:

`toolbar.add_heading=.....`

`toolbar.list_ordered=.....`

or 

`toolbar.add_heading.tooltip=.....`

File: web_src/css/base.css
Comment (delvh) [reply web_src/css/base.css]: Well… My definition of `muted` is that the element it is applied to should be stripped of its default appearance.
That means for a link that it is displayed in the default text color, or for a button that it hasn't this visual distinction marking it immediately as a button.
For example, the `>` on the `Changed File of a PR` tab is for me a prime example that should be a button but **should not** be immediately highlighted.

Comment (silverwind) [reply web_src/css/base.css]: I see muting as "remove color unless hovered". Because color in UI "screams" for attention. It only makes sense for stuff that has color in first place, e.g. only `<a>`.

File: templates/shared/combomarkdowneditor.tmpl
Comment (delvh) [inline templates/shared/combomarkdowneditor.tmpl]: @silverwind 
> If you make the parent a flexbox (gt-df), the whitespace between tags will not matter anymore.

This is already the case, and yet the whitespace mattered…

File: web_src/css/base.css
Comment (wxiaoguang) [reply web_src/css/base.css]: I guess we do not need to introduce `button.muted` here.

Instead, we had better to have `markdown-toolbar button`, because, in the future, we need to add hover styles for these markdown toolbar elements.

The `button.muted` isn't general enough at the moment.

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Hmm can't be as per https://www.w3.org/TR/css-flexbox-1/#flex-items

> if the entire sequence of child [text runs](https://www.w3.org/TR/css-display-3/#css-text-run) contains only [white space](https://www.w3.org/TR/CSS2/text.html#white-space-prop) (i.e. characters that can be affected by the [white-space](https://www.w3.org/TR/css-text-3/#propdef-white-space) property) it is instead not rendered (just as if its [text nodes](https://www.w3.org/TR/css-display-3/#css-text-node) were display:none).

I guess I need to test it.

Comment (wxiaoguang) [reply templates/shared/combomarkdowneditor.tmpl]: Have you tried to make `markdown-toolbar-group` have `display: inline-flex`?

File: web_src/css/base.css
Comment (delvh) [reply web_src/css/base.css]: > Because color in UI "screams" for attention

Agreed. And so does displaying something as a (raw) button.
So, here's a new attempt to define `muted`:
Remove the attention an element generates through its default appearance.
That's what I've tried to state above but have struggled with.

---

> The `button.muted` isn't general enough at the moment.

(In which way are you even arguing here? Should the selector only be `muted` without the `button`, or should it be more specific?)
Either way, it is **supposed** to be like that.
Anything more general probably won't work as there are still differences between what it should do for `links` and `button`.
Anything more specific is not a good idea for something that is supposed to be a helper class to make it easy to remove the default appearance of a button.

> in the future, we need to add hover styles for these markdown toolbar elements

Why should this not be possible?
It isn't `!important`, so the pseudo-class selector will have a higher specificity and overwrite the `muted` class…

Comment (wxiaoguang) [reply web_src/css/base.css]: > > in the future, we need to add hover styles for these markdown toolbar elements
> 
> Why should this not be possible? It isn't `!important`, so the pseudo-class selector will have a higher specificity and overwrite the `muted` class…

What's your plan of doing so? Usually toolbar buttons should have visual feedbacks (eg: GitHub's toolbar buttons have hover styles)

If you do:

```
// in base.css
button.muted { ... }

// in editor-markdown.css
markdown-toolbar button:hover {}

// then
<markdown-toolbar>
    <button class="muted">...</button>
<markdown-toolbar>
```

How about just do this instead:

```
// in editor-markdown.css
markdown-toolbar button {}
markdown-toolbar button:hover {}

// then
<markdown-toolbar>
    <button>...</button>
<markdown-toolbar>
```


I believe the second one is clearer and it's easier to be fine-tuned for various cases.


(That's what I meant "I guess we do not need to introduce `button.muted` here / it isn't general enough at the moment" in https://github.com/go-gitea/gitea/pull/23896#discussion_r1156159805 )

Comment (delvh) [reply web_src/css/base.css]: Yes, depending on what you want to achieve.
As I mentioned, I'm trying to make `muted` a new helper class for buttons:
If it was only for this feature I'd agree that we do not need the class and can keep it inside the editor CSS file.

However, more instances require it:
> For example, the `>` (Expand/Collapse) on the `Changed File of a PR` tab is for me a prime example that should be a button with the `muted` class.

> This will help reduce a11y infractions in the future, as that was one of the major points why people didn't use `<button>` tags and decided on a bad fix (i.e. through `<div>`s) instead.

That's why I think it is better to make it a global helper class that behaves the same for all buttons.

File: options/locale/locale_en-US.ini
Comment (delvh) [reply options/locale/locale_en-US.ini]: What about something like `[editor].buttons.heading.tooltip`?

File: web_src/css/base.css
Comment (wxiaoguang) [reply web_src/css/base.css]: TBH, I also need such buttons, and it already exists in Fomantic library: `ui button tertiary`.  The oauth key generation / repo home topic manage buttons are all `ui button tertiary`.

So, I think we had better have a unique button definition for these cases, instead of `muted`, I prefer to describe a kind of element by its function and purpose, instead of using antonym words for them (eg: what styles/attributes/behaviors are "muted"?).

File: options/locale/locale_en-US.ini
Comment (wxiaoguang) [reply options/locale/locale_en-US.ini]: Although I prefer "toolbar.xxx", I have no objection for either.

My thought is: in the "editor" component, there could be many other buttons, for example, "link" button (add a link, or link to another issue/PR?), "preview" tab/button, "save" button (save draft?),  etc. Normal buttons are usually different from toolbar buttons and these buttons may have different titles / tooltips.

But it's not that important, either is fine.

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [inline templates/shared/combomarkdowneditor.tmpl]:
```suggestion
		<markdown-toolbar class="gt-df gt-ac gt-gap-3">
```

image

There is a noticeable gap between the element groups, I like it.

Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Not sure what the whitespace problem was but see my suggestion below, it looks alright to me.

File: web_src/css/base.css
Comment (silverwind) [reply web_src/css/base.css]: These buttons need a hover primary color.

We do need a class for buttons that look like links and which does only the color addition, no opinionated size styling like fomantic does. I suggest `btn-link` that does not depend on fomantic. Github calls this `btn-octicon`, but I think we can be more generic.

```css
.btn-link:hover {
  color: var(--color-primary);
}
```

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
code.tooltip = Add code
```

File: templates/shared/combomarkdowneditor.tmpl
Comment (wxiaoguang) [reply templates/shared/combomarkdowneditor.tmpl]: If I understand correctly, delvh wants to remove the spaces between the toolbar buttons.

```html
<markdown-toolbar class="gt-df">
	<div class="markdown-toolbar-group">
		<md-bold class="markdown-toolbar-button">...</md-bold><!-- SPACE -->
	</div>
</markdown-toolbar>
```

To hide the SPACE, its direct parent element should have `display: flex/inline-flex`.

The question in the comment: the `<markdown-toolbar class="gt-df">` only affects its direct children: `<div class="markdown-toolbar-group">`, so the SPACE is still there.

Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Indeed, flexbox space removal only affects direct children.

**Comment by silverwind** [comment]
Would you mind adding `role="button"` to all the button elements here, except `markdown-switch-easymde` (will do separately)?

File: options/locale/locale_en-US.ini
Comment (delvh) [reply options/locale/locale_en-US.ini]: The problem I can see with that:
I think `code` is surprisingly unspecific.
Especially since it is basically just text highlighting.
When I read it myself, I have no idea what the result of clicking that button will be, thus making the tooltip useless.
What about the opposite?
```suggestion
code.tooltip = Add highlighted text
```

Comment (silverwind) [reply options/locale/locale_en-US.ini]: It will result in the text wrapped in `<code>`, so it seemed descriptive enough to me. Highlight seems incorrect as that sounds to me like syntax highlighting, which this isn't.

File: web_src/css/base.css
Comment (silverwind) [inline web_src/css/base.css]: I did a simliar style in https://github.com/go-gitea/gitea/pull/24034/files#diff-fa20dee2f623fb1b315245623cdd47e3be2f346cbbe5712825f6aeb6e92d7912, but specific to these buttons. I think we should remove this style and you copy the rules from that PR here (or wait until it lands).

Comment (delvh) [reply web_src/css/base.css]: I don't like to repeat myself, but that's why I said
> The effect of the muted class is what you would expect: The button loses all of its normal styling, and is defined only by its content instead.
This will help reduce a11y infractions in the future, as that was one of the major points why people didn't use <button> tags and decided on a bad fix (i.e. through <div>s) instead.

We need a class like that, no matter how it is called.
There is no way around that.
Everything else is simply not maintainable for something as large as Gitea

**Comment by delvh** [comment]
> Would you mind adding `role="button"` to all the button elements here, except `markdown-switch-easymde` (will do separately)?

Hmm… Isn't that effectively the responsibility of the `md-*` elements to declare themselves as buttons unless otherwise specified?

**Comment by silverwind** [comment]
Changes done, this is the hover:

Screenshot 2023-04-11 at 00 15 20

File: web_src/css/base.css
Comment (silverwind) [reply web_src/css/base.css]: I agree we need a class, but I don't agree on the muted name. I would call it `btn-link`.

File: web_src/css/base.css
Comment (silverwind) [reply web_src/css/base.css]: Let's move to thread below, this one is getting too long.

File: options/locale/locale_en-US.ini
Comment (silverwind) [reply options/locale/locale_en-US.ini]: Don't you agree that essentially all stuff in backticks is "code"?

File: web_src/css/base.css
Comment (silverwind) [reply web_src/css/base.css]: I'll push up something, it combines both `btn-link` (making button like a link) and `muted` class (removing  color unless hovered).

Comment (silverwind) [reply web_src/css/base.css]: Pushed

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
The last style addition makes `btn-link` completely behave like a link. The `text-decoration` has no effect on SVG icons as they contain no text. I also copied the `a:hover { text-decoration: underline}` selector from Fomantic CSS into our CSS, to show these two styles belong together.

**Comment by delvh** [comment]
ping for other maintainers. All previous concerns have been addressed by both @silverwind and me.

**Review by lunny — APPROVED** [review]

File: templates/shared/combomarkdowneditor.tmpl
Comment (wxiaoguang) [inline templates/shared/combomarkdowneditor.tmpl:24]: I really dislike this `btn-link muted`, it just duplicates a lot.

Why can't we put the styles in `.markdown-toolbar-button` ?

Comment (wxiaoguang) [reply templates/shared/combomarkdowneditor.tmpl:24]: I won't say blocker for this change, but I will refactor such usage in next PR.

I do not see why we should add a CSS class then mute some of the styles, and I do not think the duplicated class names are maintainable.


As I explained above, we are styling the toolbar buttons, we only need a `.markdown-toolbar-button {}` style for them.

Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl:24]: > Why can't we put the styles in `.markdown-toolbar-button` ?

That was my initial approach as well but then @delvh requested more generic classes to achieve the effect. I'm indifferrent, but those classes do might prove useful in the future.

**Pull request merged**
